### PR TITLE
feat(ui): show unhandled errors on the ui

### LINKED ISSUE #4317: Show unhandled errors on the UI
### Clear and concise description of the problem

I was using the Vitest UI to monitor tests while working on an app. I didn't realize that the suite was failing, because of an unhandled error, up until I've looked at the console. I would love, if in case of an unhandled error, Vitest UI told me that my suite is failing. 

I would be happy to work on this feature.

### Suggested solution

If the suite/file where the error was thrown is always known, then display the error in the results pane of that suite.

### Alternative

_No response_

### Additional context

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by spiroka**
### Description

Adds info about unhandled errors to the ui and html reports.

Screenshot 2023-12-15 at 21 08 39
Screenshot 2023-12-15 at 21 08 28

closes #4317 

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [x] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

**Comment by dammy001** [comment]
@spiroka this ready for review?

**Comment by spiroka** [comment]
@dammy001
I've added tests. The PR is ready for review.

File: packages/ui/client/composables/client/index.ts [hunk 1]
```diff
@@ -75,6 +77,7 @@ watch(
         client.rpc.getConfig(),
       ])
       client.state.collectFiles(files)
+      unhandledErrors.value = await client.rpc.getUnhandledErrors()
```
Comment (sheremet-va) [inline packages/ui/client/composables/client/index.ts:80]: Should we put it inside the `Promise.all` a few lines earlier?

Comment (spiroka) [reply packages/ui/client/composables/client/index.ts:80]: ohh yeah, will do

Comment (spiroka) [reply packages/ui/client/composables/client/index.ts]: done

File: packages/ui/client/components/dashboard/TestFilesEntry.vue [hunk 2]
```diff
@@ -44,12 +44,33 @@ import { filesFailed, filesSnapshotFailed, filesSuccess, time } from '../../comp
...
     </div>
   </div>
+  <template v-if="unhandledErrors.length">
+    <div bg="red500/10" text="red500" p="x3 y2" max-w-xl m-2 rounded>
+      <h3 text-center mb-2>
+        Unhandled Errors
+      </h3>
+      <p text="sm" font-thin data-testid="unhandled-errors">
+        Vitest caught {{ unhandledErrors.length }} error{{ unhandledErrors.length > 1 ? 's' : '' }} during the test run.<br>
+        This might cause false positive tests. Resolve unhandled errors to make sure your tests are not affected.
```
Comment (sheremet-va) [inline packages/ui/client/components/dashboard/TestFilesEntry.vue:70]: Shouldn't we display errors somehow? (A new tab? Hideable list?) This might be a bit more useful

File: packages/ui/client/components/dashboard/ErrorEntry.vue [hunk 3]
```diff
@@ -0,0 +1,60 @@
+<script setup lang="ts">
+  import { isPrimitive } from 'vitest/utils'
```
Comment (sheremet-va) [inline packages/ui/client/components/dashboard/ErrorEntry.vue:2]: Please don't import from `vitest/` in UI package. Only relative paths are allowed there. This particular function can just be inlined

File: packages/ui/client/components/dashboard/TestFilesEntry.vue (hunk 2 above)
Comment (spiroka) [reply packages/ui/client/components/dashboard/TestFilesEntry.vue:70]: I've added some details about the errors. 
Took [this](https://github.com/vitest-dev/vitest/blob/main/packages/vitest/src/node/error.ts#L25) method as a starting point. I've kept things simple for now, for example it doesn't display the stack trace. Please let me know if you think I should add more info.

Also test cases do not cover every branch yet, as I was not yet able to come up with a sample test for all the edge cases.

Comment (sheremet-va) [reply packages/ui/client/components/dashboard/TestFilesEntry.vue:70]: We should definitely print stack trace, maybe without the code block, but stack should be present

Comment (spiroka) [reply packages/ui/client/components/dashboard/TestFilesEntry.vue:70]: Ok, will do

File: packages/ui/client/components/dashboard/ErrorEntry.vue [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    This error originated in <span font-bold>{{ e.VITEST_TEST_PATH }}</span> test file. It doesn't mean the error was thrown inside the file itself, but while it was running.
+  </p>
+  <p v-if="e.VITEST_TEST_NAME" text="sm" font-thin mb-2>
+    The latest test that might've caused the error is <span font-bold>{{ e.VITEST_TEST_NAME }}</span>. It might mean one of the following:<br>
+    <ul>
+      <li>
+        The error was thrown, while Vitest was running this test.
+      </li>
+      <li>
+        This was the last recorded test before the error was thrown, if error originated after test finished its execution.
```
Comment (dammy001) [inline packages/ui/client/components/dashboard/ErrorEntry.vue:47]:
```suggestion
        If the error occurred after the test had been completed, this was the last documented test before it was thrown.
```

Comment (dammy001) [reply packages/ui/client/components/dashboard/ErrorEntry.vue:47]: @spiroka did some rewording here.

File: packages/ui/client/components/dashboard/ErrorEntry.vue [hunk 5]
```diff
@@ -0,0 +1,62 @@
+<script setup lang="ts">
+  function isPrimitive(value: unknown) {
+    return value === null || (typeof value !== 'function' && typeof value !== 'object')
+  }
+
+  const props = defineProps<{
+    error: unkown
```
[the remark this answers is not in the document]
Comment (spiroka) [reply packages/ui/client/components/dashboard/ErrorEntry.vue:7]: oh wow, how did that even work

File: packages/ui/client/components/dashboard/ErrorEntry.vue (hunk 4 above)
Comment (spiroka) [reply packages/ui/client/components/dashboard/ErrorEntry.vue:47]: @dammy001 Thanks! I will update this text for the CLI as well then.

File: packages/ui/client/components/dashboard/TestFilesEntry.vue (hunk 2 above)
Comment (spiroka) [reply packages/ui/client/components/dashboard/TestFilesEntry.vue:70]: @sheremet-va Added stack trace to errors. Updated the description with new screenshots.

**Comment by sheremet-va** [comment]
Can you resolve a conflict please?

**Comment by spiroka** [comment]
> Can you resolve a conflict please?

done

**Review by sheremet-va — APPROVED** [review]

**Review by dammy001 — APPROVED** [review]

File: packages/ui/client/composables/error.ts [hunk 6]
```diff
@@ -1,4 +1,6 @@
 import Filter from 'ansi-to-html'
+import type { ErrorWithDiff } from 'vitest'
+import { parseStacktrace } from '../../../utils/src/source-map'
```
Comment (hi-ogawa) [inline packages/ui/client/composables/error.ts:3]: Should this be imported from `@vitest/utils/source-map`? Currently this is directly referencing typescript file in `packages/utils/src/source-map.ts`.

Probably client js will be bundled in the same way, but it might be odd that since `packages/ui` has `"@vitest/utils": "workspace:*"` dependency explicitly.

Comment (sheremet-va) [reply packages/ui/client/composables/error.ts:3]: Whole UI is bundled so it doesn't really matter

Comment (sheremet-va) [reply packages/ui/client/composables/error.ts:3]: Interestingly enough, changing to `@vitest/utils` makes bundler a bit smaller
